lib/WTSI/NPG/iRODS/icommands.pm

```diff
@@ -0,0 +1,109 @@
+package WTSI::NPG::iRODS::icommands;
+
+use strict;
+use warnings;
+
+use English qw(-no_match_vars);
+use Exporter qw[import];
+our @EXPORT_OK = qw[iquest];
+
+our $VERSION = '';
+
+my $log = Log::Log4perl->get_logger(__PACKAGE__);
+
+=head2 iquest
+
+  Arg [1]    : iquest arguments
+
+  Example    : my @result = iquest('-z', $zone, q[%s],
+                              "select COLL_NAME where COLL_NAME like '$coll%'");
+  Description: Return results of an iquest query. The arguments passed to iquest
+               should be strings as used on the command line. The exception to
+               this is that the '--no-page' argument may be omitted as this
+               function always fetches all results without paging. It is the
+               caller's responsibility to ensure correct quoting within the
+               query string. If a query finds no results, an empty array is
+               returned.
+  Returntype : Array of result strings
+
+=cut
+sub iquest {
+  my (@args) = @_;
+
+  @args = grep { $_ ne '--no-page' } @args;
+
+  my @iquest_cmd = ('iquest', '--no-page', @args);
+  my $cmd = join q[ ], @iquest_cmd;
+
+  $log->debug("Executing '$cmd'");
+  open my $fh, q[-|], @iquest_cmd or
+    $log->logcroak("Failed to open pipe from '$cmd': $ERRNO");
+
+  my @records;
+  while (my $line = <$fh>) {
+    chomp $line;
+    next if $line =~ m{^\s*$}msx;
+
+    # Work around the iquest bug/misfeature where it mixes its logging output
+    # with its data output
+    next if $line =~ m{^Zone is}msx;
+    next if $line =~ m{^CAT_NO_ROWS_FOUND}msx;
+    # Skip record separators. These are not printed consistently; when there
+    # are sufficient records to cause pagination, the separator is missing.
+    next if $line =~ m{^----$}msx;
+
+    $log->debug("iquest: $line");
+    push @records, $line;
+  }
+
+  my $rc = close $fh;
+  my $exit_code = $CHILD_ERROR >> 8;
+  if ($exit_code == 0) {
+    return @records;
+  }
+  if ($exit_code == 1) {
+    $log->debug("Empty result from '$cmd'");
+    return @records;
```

Comment:
It looks as though iquest also exits 1 if there is no irods environment (https://github.com/irods/irods_client_icommands/blob/72b50d16397675eacc63c577772e7a6da8f67452/src/iquest.cpp#L377). This could cause confusion... perhaps you could set a flag when skipping lines with `NO_ROWS_FOUND`?

Reply:
Oh, that looks like a bug. they can't have a special meaning for `1` (success with no results) and also use it for an error.

Reply:
That needs fixing in `iquest`.

Reply:
Is a fix likely to be backported to the version we move to next?  If not, we should have some sort of workaround that can be removed once we move to a fixed version.

Reply:
In practice, though it's hard to trigger the failure in `getRodsEnv` which is needed to get the incorrect code. All the invalid files I've tried so far silently pass `getRodsEnv` even though it fails to do anything useful with the file, but then `iquest` fails later with e.g. unknown hostname. That means `iquest` doesn't exit `1` in the end.

Reply:
So I think `getRodsEnv` _should_ fail and we _should_ see an incorrect exit code, but the one bug masks the other.